### ISSUE TITLE
feat: giving the director a "notepad" for inter-turn directorial memory

### ISSUE THREAD

**Issue by hpnyaggerman** [OrbFrontend/Orb#4]

**Comment by OrbFrontend** [comment]
A note taking tool with an expiration on the notes sounds good.

My only doubt is how well the model will make use of this tool. Maybe we'll have a harness to force this tool for a long-term direction every few turns instead of saying "Hey you have this notepad, use it however you want." This is because our targets are 10-30B class models here, not something like Claude Opus.

**Comment by hpnyaggerman** [comment]
Very adjacent to what was implemented in https://github.com/OrbFrontend/Orb/pull/50, pretty sure this can be implemented either as a workflow in some form and/or by expanding the surface of workflows.

**Comment by hpnyaggerman** [comment]
@OrbFrontend Cooking something up on this, if you can delay creating new migrations or altering things around the director too much, then that would be great.

**Comment by OrbFrontend** [comment]
Not much on my backlog at the moment for Orb, probably. My next target is the text completion mode, which will probably take months to properly think and design around, can't rush it.

<img width="754" height="741" alt="Image" src="https://github.com/user-attachments/assets/3830854f-e54e-41dc-bc36-51bf3cca5875" />

I'm working on something adjacent in the meantime. A 68M param slop classifier. 7ms per sentence on AM4 DDR4 hardware. Most of the dataset was manually annotated by me, it has ~90% accuracy on slop classification task.

My vision is to have something as close to inpainting as possible for text. One pass, then several more afterwards for a perfect final output, with no latency wasted. Text Completion is one piece, classifiers are another piece.

**Comment by hpnyaggerman** [comment]
@OrbFrontend In regards to detecting and fixing purple prose. Have you considered reducing the sentences to their most basic semantic representation and then rewriting them in non-purple-prose-y way? On that note, if one decomposes purple prose ridden LLM-generated sentences into their most basic semantic representation, one is likely to find that many of them can be eliminated without losing any meaning. I think this is worth considering as a vector of inquiry, also.

**Comment by hpnyaggerman** [comment]
@OrbFrontend Also you might want to look into https://github.com/p-e-w/heretic. It is made for uncensoring ablation but can be used for anti-slop ablation also.

**Comment by hpnyaggerman** [comment]
@OrbFrontend So what's the progress on the de-purple-prosing? I've seen a few posts you've made on lmg but so far no source code. Will you release it under OrbFrontent or a different account? I think OrbFrontend would make sense (you've already referred to yourself as orbanon and to orb as your project in the context of this new one). Regardless, would be interested in seeing and seeing how it and possibly making improvements upon it because I have worked on heretic before and may have ways of making the ablation process more smooth and deeply penetrating.

**Comment by OrbFrontend** [comment]
@hpnyaggerman I'm still working on it, needs some tidying, I'm also doing a write-up... Will release it under this same account soon, I'm juggling several accounts already so no need to add another. Will probably need help with the dataset, it's serviceable but bad, AI-generated pairs (the core of the ablation) are one-directional and not very diverse in their style and word choices.

**Comment by hpnyaggerman** [comment]
@OrbFrontend I am currently working on my own dataset for heretic. Well, more of a module with an embedded dataset. One the purpose of which is to eval both refusal and quality in one go and in a deterministic manner rather than relying on string-matching refusal detection. But once I am done with that and once you publish the repo with the code, I would potentially be interested in helping, as ablating models in multiple modes simultaneously (safety un-alignment + prose de-purpling) sounds like it would produce novel and potentially useful results.

**Comment by OrbFrontend** [comment]
@hpnyaggerman Still struggling with the data. My approach was wrong. Rather than trying to remove "purple/vivid prose", I should focus on common AI slop instead. LLM outputs from the past three years have consistent, homogeneous patterns and can be identified easily. And rather than writing everything to be more terse, I'll rewrite into more something interesting, still wonder how this will impact the direction vector. The "interesting" part sounds flaky but it's worth a try.

Gonna take some more time until a release. Moving from de-prose to de-slop.

**Comment by hpnyaggerman** [comment]
@OrbFrontend Here is an idea for you. Hook the purple classifier up to heretic and use it as a source of scoring for trials. This way you don't even need to think about what the right data is. Heretics ARA branch does not (as far as I remember) require you to get any kind of an initial direction by inferring against positive and negative prompts. And their WIP but soon to be released plugin system allows arbitrary ablation objectives.

I am sure you've had this idea but in case you haven't, well there you go.

And in regards to what to prompt the model for during eval, you can have both one/zero-shot creative writing prompts and/or you can also look around on 4chan for logs of roleplay sessions, human-to-human or human-to-llm, and tell the LLM to continue those, and classify the response and rank it based on that. That's what I'd do probably.

**Comment by OrbFrontend** [comment]
@hpnyaggerman The classifier is actually the ground for scoring the trials, then cheating penalties (taken from Orb detectors) are added to the scores.

Data-wise, that's more or less what I tried with the 31B run, there were no slop samples in the direction at all, only overwroughtly rewritten wiki/news pairs. The optimizer simply optimized for the lowest score, sorta like unsupervised training. Just wanted to see what would happen. The prose was ablated away, short-form slop like "Her voice dripped with disdain" stayed because it was terse and there were no negs for it. The bait scenarios are actually diverse multi-turn RP that will draw out the slop specified in the directions (~20 premeditated scenarios, but no zero-shots), purple prose and repetition emerge over time.

The implementation is just simple orthogonalization. Tbh I don't understand ARA (yet), it gives me unsupervised feels, but first I'm trying to experiment whether I can somehow change the writing style to my liking instead of letting the model arbitrarily veer wherever it wants as long as it's not caught by classifier cop. Sorta like poor-man's SFT, probably sounds stupid but I want to try. So pos/neg directions are what I'm experimenting on. Still, I'd like to see how ARA works.

This is the write-up I'm doing, it's unfinished but the core ideas are there: [write_up.md](https://github.com/user-attachments/files/29610526/write_up.md).

There are still several things I want to try and several mistakes I plan to fix. But maybe I'll just upload it unfinished and see if anyone has any better idea.

**Comment by hpnyaggerman** [comment]
@OrbFrontend I think it would be a good idea to publish the repo sooner rather than later and then just do the work you are interested in doing in branches dedicated to it. I am sure you will find interest in this project among anons and not only anons. I have a few thoughts of my own but I would have to complete my own work first and then wait for heretic to finally merge the plugin system.

**Comment by hpnyaggerman** [comment]
@OrbFrontend fround this https://github.com/sam-paech/slop-forensics, thought you might be interested

**Comment by OrbFrontend** [comment]
> [@OrbFrontend](https://github.com/OrbFrontend) fround this https://github.com/sam-paech/slop-forensics, thought you might be interested

I saw this a few years ago. Might good for automatically generating prominent slop for training data...

**Comment by OrbFrontend** [comment]
@hpnyaggerman I have another idea to use GRPO where the analyzers and the classifier act as reward functions, probably need to train the classifier on dialogue slop, and add a dialogue percentage gate to prevent cheating by turning everything into a dialogue. The direct rank-1 cutting approach is too blind for _slop_ ablation and I've been failing at generate satisfying synth data for directed ablation. Not sure if 24GB VRAM is enough to experiment on E4B...

**Comment by OrbFrontend** [comment]
Gonna close this issue since the feature has been implemented.
